streams/src/main/java/org/apache/kafka/streams/processor/internals/assignment/KafkaStreamsStateImpl.java

```diff
@@ -0,0 +1,140 @@
+import static java.util.Comparator.comparingLong;
+
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+import java.util.SortedMap;
+import java.util.SortedSet;
+import java.util.TreeSet;
+import java.util.stream.Collectors;
+import org.apache.kafka.streams.processor.TaskId;
+import org.apache.kafka.streams.processor.assignment.KafkaStreamsState;
+import org.apache.kafka.streams.processor.assignment.ProcessId;
+import org.apache.kafka.streams.state.HostInfo;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class KafkaStreamsStateImpl implements KafkaStreamsState {
+    private static final Logger LOG = LoggerFactory.getLogger(KafkaStreamsStateImpl.class);
+
+    private final ProcessId processId;
+    private final int numProcessingThreads;
+    private final Map<String, String> clientTags;
+    private final Map<TaskId, Long> taskLagTotals; // contains lag for all stateful tasks in the app topology
+    private final SortedSet<TaskId> previousActiveTasks;
+    private final SortedSet<TaskId> previousStandbyTasks;
+    private final SortedMap<String, Set<TaskId>> taskIdsByConsumer;
+    private final Optional<HostInfo> hostInfo;
+
+    public KafkaStreamsStateImpl(
+        final ProcessId processId,
+        final int numProcessingThreads,
+        final Map<String, String> clientTags,
+        final Map<TaskId, Long> taskLagTotals,
+        final SortedSet<TaskId> previousActiveTasks,
+        final SortedSet<TaskId> previousStandbyTasks,
+        final SortedMap<String, Set<TaskId>> taskIdsByConsumer,
+        final Optional<HostInfo> hostInfo
+    ) {
+        this.processId = processId;
+        this.numProcessingThreads = numProcessingThreads;
+        this.clientTags = unmodifiableMap(clientTags);
+        this.taskLagTotals = unmodifiableMap(taskLagTotals);
+        this.previousActiveTasks = unmodifiableSortedSet(previousActiveTasks);
+        this.previousStandbyTasks = unmodifiableSortedSet(previousStandbyTasks);
+        this.taskIdsByConsumer = unmodifiableSortedMap(taskIdsByConsumer);
+        this.hostInfo = hostInfo;
+    }
+
+    @Override
+    public ProcessId processId() {
+        return processId;
+    }
+
+    @Override
+    public int numProcessingThreads() {
+        return numProcessingThreads;
+    }
+
+    @Override
+    public SortedSet<String> consumerClientIds() {
+        return new TreeSet<>(taskIdsByConsumer.keySet());
+    }
+
+    @Override
+    public SortedSet<TaskId> previousActiveTasks() {
+        return previousActiveTasks;
+    }
+
+    @Override
+    public SortedSet<TaskId> previousStandbyTasks() {
+        return previousStandbyTasks;
+    }
+
+    @Override
+    public long lagFor(final TaskId task) {
+        final Long totalLag = taskLagTotals.get(task);
+        if (totalLag == null) {
+            throw new IllegalStateException("Tried to lookup lag for unknown task " + task);
```

Comment:
As a general rule that we clearly don't always follow in Streams, but should/are actively getting better about, we should make sure to always log an error before throwing a (new) exception. Sometimes the error handling can make it difficult to trace back an exception to the original source of error, and error logs help with that. Doesn't need to be too complicated -- although in this case, perhaps it would make sense to include the keySet of `taskLagTotals` in the error log (I'm thinking to help differentiate the case where this specific task isn't included vs the entire map being empty)